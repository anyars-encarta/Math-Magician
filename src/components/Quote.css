*,
body {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font-family:
    -apple-system,
    BlinkMacSystemFont,
    'Segoe UI',
    'Roboto',
    'Oxygen',
    'Ubuntu',
    'Cantarell',
    'Fira Sans',
    'Droid Sans',
    'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.quote-container {
  padding: 10px;
}

h1 {
  margin-bottom: 20px;
}

h1,
.quote,
.author-category,
.loading,
.error {
  font-size: 30px;
  text-align: center;
}

.loading {
  color: blue;
}

.quote,
.error {
  color: red;
}

.quote {
  font-style: italic;
  margin-bottom: 30px;
}

hr {
  margin: 0 auto;
  width: 300px;
}
