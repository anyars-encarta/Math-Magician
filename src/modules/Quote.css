*,
body {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font-family:
    -apple-system,
    BlinkMacSystemFont,
    'Segoe UI',
    'Roboto',
    'Oxygen',
    'Ubuntu',
    'Cantarell',
    'Fira Sans',
    'Droid Sans',
    'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.quote-container {
  padding: 10px;
  max-width: 60%;
  margin: 0 auto;
  margin-top: 50px;
  background-color: rgba(126, 119, 119, 0.816);
}

.quote-header {
  margin-bottom: 20px;
  color: #ffff;
  text-shadow: 1px 1px 2px red, 0 0 1em blue, 0 0 0.2em blue;
}

h1,
.quote,
.author-category,
.loading,
.error {
  font-size: 30px;
  text-align: center;
}

.loading {
  color: blue;
}

.error {
  color: red;
}

.author-category {
  color: rgb(145, 255, 0);
}

.quote {
  font-style: italic;
  margin-bottom: 30px;
}

hr {
  margin: 0 auto;
  width: 300px;
}
