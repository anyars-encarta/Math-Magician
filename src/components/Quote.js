import React, { useEffect, useState } from 'react';
import './Quote.css';

const uniqueID = '8YHMipVLjkYiF4KOWb7GUYqkO6VXXgcUK3qQJEAI';

function Quote() {
  const apiUrl = 'https://api.api-ninjas.com/v1/quotes';
  const [quoteText, setQuoteText] = useState('');
  const [authorText, setAuthor] = useState('');
  const [categoryText, setCategory] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const [hasError, setHasError] = useState(false);

  useEffect(() => {
    const fetchQuote = async () => {
      try {
        const response = await fetch(apiUrl, {
          headers: {
            'X-Api-Key': uniqueID,
          },
        });
        if (!response.ok) {
          throw new Error(
            `API request failed with status code ${response.status}.`,
          );
        }
        const data = await response.json();
        setQuoteText(data[0].quote);
        setAuthor(data[0].author);
        setCategory(data[0].category);
      } catch (error) {
        setHasError(true);
      }
      setIsLoading(false);
    };

    fetchQuote();
  }, []); // Empty dependency array to run only once on initial component mount

  const fetchQuote = async () => {
    setIsLoading(true);
    try {
      const response = await fetch(apiUrl, {
        headers: {
          'X-Api-Key': uniqueID,
        },
      });
      if (!response.ok) {
        throw new Error(
          `API request failed with status code ${response.status}.`,
        );
      }
      const data = await response.json();
      setQuoteText(data[0].quote);
      setAuthor(data[0].author);
      setCategory(data[0].category);
    } catch (error) {
      setHasError(true);
    }
    setIsLoading(false);
  };

  useEffect(() => {
    const timer = setInterval(fetchQuote, 30000); // Change quote every 30 seconds

    return () => clearInterval(timer); // Clear the timer on component unmount
  }, []); // Empty dependency array to run only once on initial component mount

  return (
    <div className="quote-container">
      <h1 className="quote-header">Encarta Maths Magician - Quote of the Day!</h1>
      {isLoading && <p className="loading">Loading Quote...</p>}
      {hasError && <p className="error">Something went wrong!</p>}

      {quoteText && (
        <p className="quote">
          &quot;
          {quoteText}
          &quot;
        </p>
      )}
      <hr />
      <br />
      {authorText && (
        <p className="author-category">
          {authorText}
          ,&nbsp;
          {categoryText}
        </p>
      )}
    </div>
  );
}

export default Quote;
