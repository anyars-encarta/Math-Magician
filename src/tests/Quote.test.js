import React from 'react';
import {
  render, waitFor, screen,
} from '@testing-library/react';
import Quote from '../components/Quote';

// Snapshot Testing
test('Quote component matches snapshot', () => {
  const { container } = render(<Quote />);
  expect(container).toMatchSnapshot();
});

// User Interaction Testing
jest.useFakeTimers();

describe('Quote component', () => {
  it('displays loading state initially', () => {
    render(<Quote />);
    expect(screen.getByText('Loading Quote...')).toBeInTheDocument();
  });

  it('displays error state when API call fails', async () => {
    global.fetch = jest.fn(() => Promise.reject('API error'));
    render(<Quote />);
    await waitFor(() => expect(screen.getByText('Something went wrong!')).toBeInTheDocument());
  });

  it('displays fetched quote and author', async () => {
    const mockData = [
      { quote: 'Mock Quote', author: 'Mock Author', category: 'Mock Category' },
    ];
    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve(mockData),
    }));

    render(<Quote />);

    await waitFor(() => expect(screen.getByText('"Mock Quote"')).toBeInTheDocument());
    expect(screen.getByText('Mock Author, Mock Category')).toBeInTheDocument();
  });
});
