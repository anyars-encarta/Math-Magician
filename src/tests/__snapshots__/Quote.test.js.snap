// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote component matches snapshot 1`] = `
<div>
  <div
    class="quote-container"
  >
    <h1
      class="quote-header"
    >
      Quotes of the Day!
    </h1>
    <p
      class="loading"
    >
      Loading Quote...
    </p>
  </div>
</div>
`;
